libevmasm/Ethdebug.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libevmasm/Ethdebug.h>
+
+using namespace solidity;
+using namespace solidity::evmasm;
+using namespace solidity::evmasm::ethdebug;
+
+namespace
+{
+
+Json programInstructions(Assembly const* _assembly, LinkerObject const* _linkerObject, unsigned _sourceId)
+{
+	// e.g. interfaces don't have a valid assembly object.
+	if (_assembly)
+	{
+		solUnimplementedAssert(_assembly->eofVersion() == std::nullopt, "ethdebug does not yet support EOF.");
+		solUnimplementedAssert(_assembly->codeSections().size() == 1, "ethdebug does not yet support multiple code-sections.");
+	}
+
+	solAssert(_linkerObject->offsetAfterLastInstruction <= _linkerObject->bytecode.size());
```

Comment:
You're apparently assuming that `_linkerObject` can't be null. Why isn't it a reference rather than a pointer?